docs/design/018-new-http-api.md

```diff
@@ -0,0 +1,402 @@
+
+  One source of inspiration is the Go `net/http` package, which the k6 team is already familiar with. Based on this, our JS API could have similar entities:
+
+  - `Dialer`: a low-level interface for configuring TCP/IP options, such as TCP timeout and keep-alive, TLS settings, DNS resolution, IP version preference, etc.
+
+  - `Transport`: interface for configuring HTTP connection options, such as proxies, HTTP version preferences, etc.
+
+    It enables advanced behaviors like intercepting requests before they're sent to the server.
+
+  - `Client`: the main entrypoint for making requests, it encompasses all of the above options. A k6 script should be able to initialize more than one `Client`, each with their separate configuration.
+
+    In order to simplify the API, the creation of a `Client` should use sane defaults for `Dialer` and `Transport`.
+
+  There should be some research into existing JS APIs that offer similar features (e.g. Node/Deno), as we want to offer an API familiar to JS developers, not necessarily Go developers.
+
+  - `Request`/`Response`: represent objects sent by the client, and received from the server. In contrast to the current API, the k6 script should be able to construct `Request` objects declaratively, and then reuse them to make multiple requests with the same (or similar) data.
+
+- All methods that perform I/O calls must be asynchronous. Now that we have `Promise`, event loop and `async`/`await` support natively in k6, there's no reason for these to be synchronous anymore.
+
+- The API should avoid any automagic behavior. That is, it should not attempt to infer desired behavior or options based on some implicit value.
+
+  We've historically had many issues with this ([#878](https://github.com/grafana/k6/issues/878), [#1185](https://github.com/grafana/k6/issues/1185)), resulting in confusion for users, and we want to avoid it in the new API. Even though we want to have sane defaults for most behavior, instead of guessing what the user wants, all behavior should be explicitly configured by the user. In cases where some behavior is ambiguous, the API should raise an error indicating so.
+
+
+#### Sockets
+
+A Socket represents the file or network socket over which client/server or peer communication happens.
+
+It can be of three types:
+- `tcp`: a stream-oriented network socket using the Transmission Control Protocol.
+- `udp`: a message-oriented network socket using the User Datagram Protocol.
+- `ipc`: a mechanism for communicating between processes on the same machine, typically using files.
+
+The Socket state can either be _active_—meaning connected for a TCP socket, bound for a UDP socket, or open for an IPC socket—, or _inactive_—meaning disconnected, unbound, or closed, respectively.
+
+##### Example
+
+- TCP:
+```javascript
+import { dialTCP } from 'k6/x/net';
+
+export default async function () {
+  const socket = await dialTCP('192.168.1.1:80', {
```

Comment:
I'm not a fan of these `dial*` functions. I added them inspired by Go's `net` package, but this doesn't provide the ability to implement custom dialers. So maybe we should expose some `Dialer` type? I wasn't sure how to structure it, though.

And I'm slightly against calling these "dialers" to begin with, since it's mostly a Go concept. At the same time, I didn't want to call this `connect()` either, since it wouldn't make sense for UDP or IPC. `dialIPC` still doesn't make sense to me, since it's essentially opening a file, but this too is following Go's `net` package. :thinking:

WDYT?

Reply:
I agree, what about trying to be consistent with the HTTP Client API and doing something similar?

We have a class, a constructor, and methods.

```js
import { TCPSocket, <Other Protocols>... } from 'k6/x/net'; // TCP, TCPSocket, TCPConn (We should use a name that makes sense for the context and is aligned with the RFC

export default async function () {
  const socket = new TCPSocket(''192.168.1.1:80'")
  await socket.open()
```

If it doesn't make sense for UDP then we can have `new UDP()` and that's all. This should also be very similar to the Linux API.

Reply:
I had something like that initially, but decided against it, since you shouldn't be able to create a socket object, without it being actually connected/bound/open. Otherwise, you might be tempted to set some of its properties, which would be meaningless unless it's actually bound to an OS socket. In fact, most/all of its properties should be read-only. This is why I opted to have a constructor function instead.

An HTTP client is a bit different, since it's only used to make requests, and doesn't correspond to some system state. So you could theoretically have a Client instance that hasn't established a socket connection yet, and making the first request would establish the connection.

I also considered something like:

```javascript
import { TCP, UDP, IPC } from 'k6/x/net';
const socket = TCP.dial('192.168.1.1:80');
```

So in a sense you have some abstract protocol object, and calling `dial()` on it returns you a `TCPSocket`, `UDPSocket` or `IPCSocket` instance. I'm slightly leaning towards this now, even though `dial` is still in the name. WDYT?

Reply:
I went with `open` in 8c97fe07e.

`connect` doesn't quite make sense for UDP or IPC, whereas `open` is generic enough to apply for all 3.

I think `dial` is not just Go's invention. It does have some networking roots, though I'm not able to find definitive references outside of Go, and "dial-up modem" :smile:, right now.

Reply:
One more thing: the proposal doesn't currently specify a way to define custom dialers.

I wonder if we could handle this with an event system as well. E.g. have `before`/`after` events scripts can subscribe to, similar to the HTTP events, emitted before/after the socket is opened.

This could be a way to handle DNS lookups, instead of the currently proposed `lookup` hook function, but it would be generic enough for any other use cases (TCP proxies, custom metrics, etc.).

So you would have something like:

```javascript
const socket = await TCP.open('myhost', {
  eventHandlers: {
    before: event => {
      event.data.address = await dns.resolve(event.data.hostname, {
        rrtype: 'A',
        servers: ['1.1.1.1:53', '8.8.8.8:53'],
      });
    },
  },
});
```

I'm not entirely sold on the syntax to change `event.data` like this, as it's unclear which properties will be used after the event handler executes. Also, event handlers typically shouldn't return any values either, as that would make them too specific for a single purpose, so returning is not an option.

I'm not really sure we should allow creating custom dialer implementations from scratch in JS. I.e. nobody would choose to implement custom protocols from scratch in JS, and if they need to do that, then our Go API should allow it by being easily extensible. So for JS, it might just be good enough to have hooks, or event handlers, so scripts can implement some custom logic around the socket implementation.